core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/internal/signer/DefaultAwsV4HttpSigner.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.auth.aws.internal.signer;
+
+import static software.amazon.awssdk.http.auth.aws.util.CredentialUtils.sanitizeCredentials;
+import static software.amazon.awssdk.http.auth.aws.util.SignerConstant.PRESIGN_URL_MAX_EXPIRATION_DURATION;
+import static software.amazon.awssdk.http.auth.spi.SignerProperty.validatedProperty;
+
+import java.time.Clock;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.function.Function;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.http.auth.aws.AwsV4HttpSigner;
+import software.amazon.awssdk.http.auth.aws.signer.Checksummer;
+import software.amazon.awssdk.http.auth.aws.signer.CredentialScope;
+import software.amazon.awssdk.http.auth.aws.signer.V4PayloadSigner;
+import software.amazon.awssdk.http.auth.aws.signer.V4Properties;
+import software.amazon.awssdk.http.auth.aws.signer.V4RequestSigner;
+import software.amazon.awssdk.http.auth.spi.AsyncSignRequest;
+import software.amazon.awssdk.http.auth.spi.AsyncSignedRequest;
+import software.amazon.awssdk.http.auth.spi.SignRequest;
+import software.amazon.awssdk.http.auth.spi.SyncSignRequest;
+import software.amazon.awssdk.http.auth.spi.SyncSignedRequest;
+import software.amazon.awssdk.identity.spi.AwsCredentialsIdentity;
+import software.amazon.awssdk.utils.Logger;
+
+/**
+ * An implementation of a {@link AwsV4HttpSigner} that uses properties to compose v4-signers in order to delegate signing of a
+ * request and payload (if applicable) accordingly.
+ */
+@SdkInternalApi
+public final class DefaultAwsV4HttpSigner implements AwsV4HttpSigner {
+
+    private static final Logger LOG = Logger.loggerFor(DefaultAwsV4HttpSigner.class);
+
+    /**
+     * Given a request with a set of properties and a base signer, compose an implementation with the base signer based on
+     * properties, and delegate the request to the composed signer.
+     */
+    private static AwsV4HttpSigner getDelegate(SignRequest<?, ? extends AwsCredentialsIdentity> signRequest) {
+        String regionName = validatedProperty(signRequest, AwsV4HttpSigner.REGION_NAME);
```

Comment:
Should we put validatedProperty as a non-static method on the signRequest?

```java 
signRequest.requireProperty(AwsV4HttpSigner.REGION_NAME)
```

I believe we have this pattern in a few places.

Reply:
I had considered this when [suggesting to move this validation to http-auth-spi](https://github.com/aws/aws-sdk-java-v2/pull/4181#discussion_r1271146086) but wasn't sure about such usage in our codebase and thought static validation method was ok. But +1 this suggestion.